### ISSUE TITLE
`needless_collect` false positive

### ISSUE THREAD

**Issue by SamRodri** [rust-lang/rust-clippy#7512]
Lint name: `needless_collect`

Reduced code:

```rust
pub struct Foo {
    foo: bool,
    bar: bool,
}

pub fn test(foo: Vec<Foo>) {
    let bar: Vec<_> = foo.iter().map(|f|f.bar).collect();
    drop(foo);

    for (_, _) in bar.into_iter().enumerate() {

    }    
}

// pub fn incorrect_suggestion(foo: Vec<Foo>) {
//     
//     drop(foo);
// 
//     for (_, _) in foo.iter().map(|f|f.bar).enumerate() {
// 
//     }    
// }
```

I expected no warning here because `foo` is dropped.  The incorrect suggestion also appears if `foo` is shadowed, potentially generating incorrect code that compiles.

### Meta

- `cargo clippy -V`:  clippy 0.1.54 (a178d032 2021-07-26)
- `rustc -Vv`:
```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-pc-windows-msvc
release: 1.54.0
LLVM version: 12.0.1
```

**Comment by jntrnr** [comment]
Just coming here to +1 this. We're having to disable this lint globally in Nushell because of the number of false positives.
